README.md
# @pixel/icons

[![Quality Gate Status](https://sonarcloud.io/api/project_badges/measure?project=cadgerfeast_pixel-icons&metric=alert_status)](https://sonarcloud.io/dashboard?id=cadgerfeast_pixel-icons)
[![Coverage](https://sonarcloud.io/api/project_badges/measure?project=cadgerfeast_pixel-icons&metric=coverage)](https://sonarcloud.io/dashboard?id=cadgerfeast_pixel-icons)
[![Version](https://badge.fury.io/js/%40pixel%2Ficons.svg)](https://www.npmjs.com/package/@pixel/icons)
[![Downloads](https://img.shields.io/npm/dt/@pixel/icons.svg)](https://www.npmjs.com/package/@pixel/icons)
[![License](https://img.shields.io/npm/l/@pixel/icons.svg)](https://github.com/cadgerfeast/pixel-icons/blob/master/LICENSE)

> Pixel icons & font sprite.

## Installation

``` bash
npm install --save @pixel/icons
```

## Quick look

<table>
	<tbody>
		<tr>
			<td align="center"><img src="./png-128/home.png"/></td>
      <td align="center"><img src="./png-128/dino.png"/></td>
      <td align="center"><img src="./png-128/floppy-disk.png"/></td>
	</tbody>
</table>

See full icon list [here.](./ICONS.md "Full icon list")

## Usage

### Font

``` html
<link rel="stylesheet" href="@pixel/icons/fonts/PixelIcons.css">
<!-- CDN -->
<link rel="stylesheet" href="https://unpkg.com/@pixel/icons/fonts/PixelIcons.css">
<!-- Usage -->
<span class="pi pi-home"></span>
```

### Icons

Icons are also available as standalone files under multiple resolutions:

* `png-16`: 16px
* `png-32`: 32px
* `png-64`: 64px
* `png-128`: 128px
* `png-256`: 256px
* `png-512`: 512px
* `png-1024`: 1024px
* `svg`: 256px

``` html
<!-- Usage -->
<img src="@pixel/icons/png-128/home.png" alt="home"/>
<img src="@pixel/icons/svg/home.svg" alt="home"/>
```

## Contributors

<table>
  <tbody>
    <tr>
      <td align="center">
        <a href="https://github.com/cadgerfeast">
          <img src="https://github.com/cadgerfeast.png?size=100" alt="cadgerfeast" width="100px">
          <br/>
          <span>cadgerfeast</span>
        </a>
      </td>
      <td align="center">
        <a href="https://github.com/zeromero-dev">
          <img src="https://github.com/zeromero-dev.png?size=100" alt="zeromero-dev" width="100px">
          <br/>
          <span>zeromero-dev</span>
        </a>
      </td>
    </tr>
  </tbody>
</table>

## License

MIT


CHANGELOG.md
# Changelog

## [0.0.14](https://github.com/cadgerfeast/pixel-icons/compare/0.0.13...0.0.14)

- migrate to 16 factor, update dependencies, add few icons [`e13a60a`](https://github.com/cadgerfeast/pixel-icons/commit/e13a60abf7b104c1ee5c7c3257d8a34e45b46407)
- fix ci, bump version 0.0.14 [`65f55c5`](https://github.com/cadgerfeast/pixel-icons/commit/65f55c5023d85e5eac98ce73ed41fb6a0ccf2358)

## [0.0.13](https://github.com/cadgerfeast/pixel-icons/compare/0.0.12...0.0.13) (2021-05-04)

- update build [`7c8bab8`](https://github.com/cadgerfeast/pixel-icons/commit/7c8bab8f590a28eec3e14b020ceaf35856e83499)
- add cross icon, bump version 0.13.0 [`273798b`](https://github.com/cadgerfeast/pixel-icons/commit/273798baba68f692f29626936d0408eb7573048f)
- Release @pixel/icons@0.0.13 [`99d7ff8`](https://github.com/cadgerfeast/pixel-icons/commit/99d7ff8b1636e3822e04462f588b3ae4a85d7649)

## [0.0.12](https://github.com/cadgerfeast/pixel-icons/compare/0.0.11...0.0.12) (2021-04-07)

- Release @pixel/icons@0.0.12 [`9cc38bb`](https://github.com/cadgerfeast/pixel-icons/commit/9cc38bba3fe45f6652e7a77ec95d1edb262ffc5c)
- re-generate fonts [`34e8bf6`](https://github.com/cadgerfeast/pixel-icons/commit/34e8bf6f4823f370ce977b8b23174a83dccb290f)
- set font size default to 30px [`2e1b895`](https://github.com/cadgerfeast/pixel-icons/commit/2e1b895ff14b07c89cb5cf1daa51b58a99f7a405)

## [0.0.11](https://github.com/cadgerfeast/pixel-icons/compare/0.0.10...0.0.11) (2021-04-07)

- bump version 0.0.11 [`f5076a7`](https://github.com/cadgerfeast/pixel-icons/commit/f5076a758d316ac1db9dd2ae4ca004f0a5fae562)
- unit test fixes [`3896226`](https://github.com/cadgerfeast/pixel-icons/commit/3896226680c4571926449bc47d2b75e89bc3c9e0)
- update build [`4903723`](https://github.com/cadgerfeast/pixel-icons/commit/49037234a08ccd01f7186a1df593e9455937faac)
- Release @pixel/icons@0.0.11 [`e83384d`](https://github.com/cadgerfeast/pixel-icons/commit/e83384df1ef7f78f4e0614ec75186d914b74a63e)

## 0.0.10 (2021-04-07)

- first draft [`4892e9c`](https://github.com/cadgerfeast/pixel-icons/commit/4892e9c34d33e3fb0d27ebafed8104eca4ea9694)
- migration to github actions [`7f787d7`](https://github.com/cadgerfeast/pixel-icons/commit/7f787d751005096ea205bbdeeeb1953c20bdf8bf)
- add release mecanism, bump version 0.0.10 [`4222770`](https://github.com/cadgerfeast/pixel-icons/commit/42227704548ac7117f053c7220c114d0ba08d7ea)
- pushing version 0.0.3, adding icons.json file [`f1d671b`](https://github.com/cadgerfeast/pixel-icons/commit/f1d671b83bd0635f03e72b55241f8063fa07ddfc)
- fixing class selector, publishing version 0.0.2 [`3ac74aa`](https://github.com/cadgerfeast/pixel-icons/commit/3ac74aa534365858310197c8e4c8990a06158194)
- add some icons, bump version 0.0.9 [`62fdc1c`](https://github.com/cadgerfeast/pixel-icons/commit/62fdc1c0283684634aa41ed22a4a3c41ff0e06fd)
- publishing version 0.0.7, adding chat and phone icons [`5077178`](https://github.com/cadgerfeast/pixel-icons/commit/5077178fc619862d46658ebbdb3c232fe6265b6e)
- publishing version 0.0.8, adding mario pipe top icon [`2f486e5`](https://github.com/cadgerfeast/pixel-icons/commit/2f486e550ae96879f9ea37e4f107c6558d211140)
- making icons darker [`bb8ba85`](https://github.com/cadgerfeast/pixel-icons/commit/bb8ba852cc29bc278c51b57a604a79b82b45ba6a)
- publishing version 0.0.5, fixing declaration file [`61a63de`](https://github.com/cadgerfeast/pixel-icons/commit/61a63de81227e43b700581311167ee6571d82c7d)
- publishing version 0.0.4, adding declaration file [`469dc9a`](https://github.com/cadgerfeast/pixel-icons/commit/469dc9a26a52664b857d9752cd795637be8de6df)
- table styling tentative [`c389958`](https://github.com/cadgerfeast/pixel-icons/commit/c389958d5d0271aeb7e3886c129ee83bd35a1964)
- shortening readme icons [`9cb7a63`](https://github.com/cadgerfeast/pixel-icons/commit/9cb7a639f4918431caab577f5524182c7dc56811)
- fixing readme links and styling [`374ab83`](https://github.com/cadgerfeast/pixel-icons/commit/374ab838261b681a7aaef59739ed2c026e4dab5e)
- Release @pixel/icons@0.0.10 [`d7622e4`](https://github.com/cadgerfeast/pixel-icons/commit/d7622e49da81ca09bda09fee300e5cfd63e7590c)
- fix coverage paths [`f344720`](https://github.com/cadgerfeast/pixel-icons/commit/f344720e1e7a31764af76962cbc602a4b2d1fe49)
- publishing version 0.0.6, migrating to a colors array [`d81db3a`](https://github.com/cadgerfeast/pixel-icons/commit/d81db3a003490b1f8c9820b55179ef2b550a752a)
- Initial commit [`c82fde2`](https://github.com/cadgerfeast/pixel-icons/commit/c82fde212dd1ca2aeb61c46592691c486ed6db4c)


build.spec.js
const svgtofont = require('svgtofont');
jest.mock('svgtofont');

describe('build', () => {
  it('should build fonts', (done) => {
    const builder = require('./build');
    builder.main().then(() => {
      expect(svgtofont).toHaveBeenCalled();
      done();
    });
  });
});


build.js
// Helpers
const fs = require('fs-extra');
const path = require('path');
const sharp = require('sharp');
const svgtofont = require('svgtofont');
// Constants
const icons = require('./icons.json');
const iconSizes = [32, 64, 128, 256, 512, 1024];
const iconPngPath = path.resolve(__dirname, 'png-16');
const iconSvgPath = path.resolve(__dirname, 'svg');
const fontsPath = path.resolve(__dirname, 'fonts');
const distPath = path.resolve(__dirname, 'dist');

let declaration = `declare interface PixelSvgData {
  svg: string;
  colors: string[];
}

declare interface PixelIcon {
  tags: string[];
  data: PixelSvgData
}

declare module '@pixel/icons' {
  const value: { [icon in PixelIconList]: PixelIcon; };
  export default value;
}`;

function rimraf(folder) {
  if (fs.existsSync(folder)) {
    fs.readdirSync(folder).forEach((entry) => {
      const currentPath = path.join(folder, entry);
      if (fs.lstatSync(currentPath).isDirectory()) {
        rimraf(currentPath);
      } else {
        fs.unlinkSync(currentPath);
      }
    });
    fs.rmdirSync(folder);
  }
}
module.exports.rimraf = rimraf;

function generateList(list) {
  let markdownContent = '# Icon list\n\n';
  markdownContent += '<table>\n\t<tbody>\n';
  markdownContent += '\t\t<tr>\n';
  let count = 0;
  for (const icon in list) {
    count++;
    markdownContent += `\t\t\t<td align="center"><img src="./png-128/${icon}.png" width="100px"/><br/><span>${icon}</span><br/><span>[${icons[icon].tags.join(', ')}]</span></td>\n`;
    if (count > 7) {
      markdownContent += '\t\t</tr>\n';
      markdownContent += '\t\t<tr>\n';
      count = 0;
    }
  }
  markdownContent += '\t</tbody>\n</table>\n';
  fs.writeFileSync(path.resolve(__dirname, 'ICONS.md'), markdownContent);
}

async function createIcon(icon) {
  const src = path.resolve(iconPngPath, icon);
  for (const size of iconSizes) {
    const destPath = path.resolve(__dirname, `png-${size}/${icon}`);
    await sharp(src).resize(size, size, { kernel: 'nearest' }).toFile(destPath);
  }
  const iSvgFilePath = path.resolve(__dirname, `svg/${icon.replace(/.png/g, '.svg')}`);
  return await generateSvg(src, iSvgFilePath);
}

async function generateSvg(src, dest) {
  const factor = Math.pow(2, 4);
  return new Promise((resolve, reject) => {
    sharp(src).raw()
      .toBuffer((err, buffer, img) => {
        if (err) { reject(err); }
        const svgData = {
          colors: []
        };
        let svgContent = `<svg xmlns="http://www.w3.org/2000/svg" shape-rendering="crispEdges" viewBox="0 0 ${img.height * factor} ${img.height * factor}" width="${img.width * factor}" height="${img.height * factor}">\n`;
        const colors = [];
        for (let i = 0; i < img.height; i++) {
          for (let j = 0; j < img.width; j++) {
            const idx = (img.width * i + j) << 2;
            if (buffer[idx + 3] !== 0) {
              const style = `rgba(${buffer[idx]},${buffer[idx + 1]},${buffer[idx + 2]},${buffer[idx + 3]})`;
              let index = colors.indexOf(style);
              if (index === -1) {
                colors.push(style);
                index = colors.indexOf(style);
              }
              svgContent += `\t<rect fill="${style}" x="${j * factor}" y="${i * factor}" width="${factor}" height="${factor}"/>\n`;
            }
          }
        }
        for (let k = 0; k < colors.length; k++) {
          svgData.colors.push(colors[k]);
        }
        svgContent += '</svg>';
        svgData.svg = svgContent;
        fs.writeFileSync(dest, svgContent);
        resolve({ svgIcon: dest, svgData: svgData });
      });
  });
}

async function main() {
  try {
    for (const size of iconSizes) {
      fs.emptyDirSync(path.resolve(__dirname, `png-${size}`));
    }
    fs.emptyDirSync(iconSvgPath);
    // Building icons
    const svgFiles = [];
    const finalIconList = {};
    const iconList = Object.keys(icons);
    const iconFiles = fs.readdirSync(iconPngPath);
    for (let i = iconFiles.length; i--;) {
      const iconName = iconFiles[i].slice(0, -4);
      if (iconList.indexOf(iconName) === -1) {
        console.warn(`${iconName} seems to not be used.`);
        iconFiles.splice(i, 1);
      } else {
        finalIconList[iconName] = icons[iconName];
      }
    }
    for (let i = iconList.length; i--;) {
      const lookout = `${iconList[i]}.png`;
      if (iconFiles.indexOf(lookout) === -1) {
        console.warn(`${iconList[i]} does not exist.`);
        iconList.splice(i, 1);
      }
    }
    declaration += '\n\ndeclare enum PixelIconList {';
    for (const icon in finalIconList) {
      const { svgIcon, svgData } = await createIcon(`${icon}.png`);
      finalIconList[icon].data = svgData;
      svgFiles.push(svgIcon);
      declaration += `\n\t'${icon}' = '${icon}',`;
    }
    declaration = declaration.slice(0, -1);
    declaration += '\n}';
    // Building icon list
    if (fs.existsSync(distPath)) {
      rimraf(distPath);
    }
    fs.mkdirSync(distPath);
    fs.writeFileSync(path.resolve(distPath, 'icons.d.ts'), declaration);
    fs.writeFileSync(path.resolve(distPath, 'icons.json'), JSON.stringify(finalIconList, null, 2));
    // Building docs
    generateList(finalIconList);
    // Building fonts
    if (!fs.existsSync(fontsPath)) {
      fs.mkdirSync(fontsPath);
    } else {
      rimraf(fontsPath);
    }
    svgtofont({
      src: iconSvgPath,
      dist: fontsPath,
      fontName: 'PixelIcons',
      classNamePrefix: 'pi',
      css: {
        fontSize: '16px'
      },
      svgicons2svgfont: {
        fontHeight: 1000,
        normalize: true
      },
      website: {
        title: 'Pixel Icons',
        logo: path.resolve(__dirname, 'svg/github.svg')
      }
    });
  } catch (error) {
    console.error(error);
  }
}
module.exports.main = main;

if (require.main === module) {
  main();
}
